cli/src/commit_templater.rs

```diff
@@ -83,6 +85,8 @@ impl<'repo> CommitTemplateLanguage<'repo> {
                 .build_cache_extensions(&mut cache_extensions);
         }
 
+        let mailmap = Rc::new(get_current_mailmap(repo, workspace_id));
```

Comment:
nit: It's probably better to cache parsed `mailmap` by caller, but this can be addressed later.

Reply:
I looked at `WorkspaceCommandHelper` and it seems like it has methods that would change the contents of the working‐copy commit? So we’d need to keep track of invalidating that state correctly, which seems like it’s maybe not worth it for a total of ~5 calls per `jj log` command that don’t seem to show up in profiling. If it ends up being a performance issue then maybe a commit ID → parsed `Mailmap` cache would be the best solution?

Reply:
This kind of cache can be managed by `ReadonlyUserRepo`. I agree the cost of loading mailmap wouldn't matter, but doing that by caller will make error handling simpler.

Reply:
I will take a look at that, thank you; I’m still getting used to the code organization.

Reply:
Could you help me understand what kind of interface you’re envisioning here? I assume it’d go in `ReadonlyRepo` as user configuration isn’t relevant here, but I’m not sure what kind of state and methods would be expected. It seems like `ReadonlyRepo` doesn’t fix a specific working‐copy commit, so I guess it’d still need to take a workspace ID as a parameter? Are you thinking that `get_current_mailmap` would move entirely to a method on `ReadonlyRepo`? Does the revset querying code have access to a `ReadonlyRepo` too?

Reply:
Ah, I meant `ReadonlyUserRepo` in cli_util.rs, which can store `OnceCell<Mailmap>` as a cache.

I think it's also good to just move `get_current_mailmap()` to `WorkspaceCommandHelper` where we can propagate load/parsing error if needed.

Reply:
Sorry for having trouble following the intent here. `ReadonlyUserRepo` doesn’t fix a specific workspace, right? So it’d have to be a `HashMap<WorkspaceId, Mailmap>` or something?

I’d definitely like to keep a helper in `jj_lib` for obtaining a `Mailmap` so that alternate frontends can easily do the right thing, per https://github.com/martinvonz/jj/pull/3951#discussion_r1653672458. But if we can figure out what the interface should look like here, it could presumably go directly in `jj_lib::repo::ReadOnlyRepo`, as the `id_prefix_context` in `ReadonlyUserRepo` isn’t relevant here and nothing about this would depend on user configuration.

Reply:
`ReadonlyUserRepo` is a private wrapper type for `WorkspaceCommandHelper` to manage cache lifecycle, and `WorkspaceCommandHelper` knows `WorkspaceId`.

So long as `Mailmap` is loaded from `@`, I don't think there is a good place to host it in `ReadonlyRepo`.